modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersValidator.java

```diff
@@ -0,0 +1,178 @@
+    public Netty4HttpHeaderValidator getValidatorInboundHandler() {
+        return new Netty4HttpHeaderValidator((httpRequest, channel, listener) -> {
+            if (httpRequest.headers() instanceof ValidatableHttpHeaders validatableHttpHeaders) {
+                // make sure validation only runs on properly wrapped "validatable" headers implementation
+                validator.apply(asHttpPreRequest(httpRequest), channel, ActionListener.wrap(validationResultContext -> {
+                    validatableHttpHeaders.markAsSuccessfullyValidated(validationResultContext);
+                    // a successful validation needs to signal to the {@link Netty4HttpHeaderValidator} to resume
+                    // forwarding the request beyond the headers part
+                    listener.onResponse(null);
+                }, e -> listener.onFailure(new HttpHeadersValidationException(e))));
+            } else {
+                listener.onFailure(new IllegalStateException("Expected a validatable request for validation"));
+            }
+        });
+    }
+
+    /**
+     * Given a {@link DefaultHttpRequest} argument, this returns a new {@link DefaultHttpRequest} instance that's identical to the
+     * passed-in one, but the headers of the latter are "validatable", in the sense that the channel handlers returned by
+     * {@link #getValidatorInboundHandler()} can use to convey the validation result.
+     */
+    public static HttpMessage wrapAsValidatableMessage(HttpMessage newlyDecodedMessage) {
+        assert newlyDecodedMessage instanceof HttpRequest;
+        DefaultHttpRequest httpRequest = (DefaultHttpRequest) newlyDecodedMessage;
+        ValidatableHttpHeaders validatableHttpHeaders = new ValidatableHttpHeaders(newlyDecodedMessage.headers());
+        return new DefaultHttpRequest(httpRequest.protocolVersion(), httpRequest.method(), httpRequest.uri(), validatableHttpHeaders);
+    }
+
+    public static ValidationResultContext extractValidationContext(org.elasticsearch.http.HttpRequest request) {
+        ValidatableHttpHeaders authenticatedHeaders = unwrapValidatableHeaders(request);
+        return authenticatedHeaders != null ? authenticatedHeaders.validationResultContextSetOnce.get() : null;
+    }
+
+    private static ValidatableHttpHeaders unwrapValidatableHeaders(org.elasticsearch.http.HttpRequest request) {
+        if (request instanceof Netty4HttpRequest == false) {
+            return null;
+        }
+        if (((Netty4HttpRequest) request).getNettyRequest().headers() instanceof ValidatableHttpHeaders == false) {
+            return null;
+        }
+        return (ValidatableHttpHeaders) (((Netty4HttpRequest) request).getNettyRequest().headers());
+    }
+
+    /**
+     * {@link HttpHeaders} implementation that carries along the {@link ValidationResultContext} iff
+     * the HTTP headers have been validated successfully.
+     */
+    public static final class ValidatableHttpHeaders extends DefaultHttpHeaders {
```

Comment:
I'm not sure I follow why we need a class here instead of just two parameters (one for the HTTPHeaders and the  other for the validation results)

Reply:
The problem that `ValidatableHttpHeaders` is solving is the one of conveying the authentication result without relying on the thread context (which is how ES authn normally works). It's the issue we discussed a couple of months ago, starting from https://github.com/elastic/elasticsearch/pull/93656/files#r1116834824 .

We intercept when the first part of a HTTP request (that cotains all the HTTP headers) is parsed by netty and we replace that instance's `HttpHeaders` member instance with this `ValidatableHttpHeaders` wrapping instance (see usages for `HttpHeadersValidator#wrapAsValidatableMessage`). The `HttpHeaders` instance is not replaced by any subsequent netty processing (in particular the `HttpObjectAggregator`) and it can also be "extracted" when the time comes to run the request handler and set the authn result back in the thread-context, as usual (see usages for `HttpHeadersValidator#extractValidationResult`).

In short, `ValidatableHttpHeaders` extends `DefaultHttpHeaders` to be able to be used in its stead, by netty, by the various types, while assembling the full http request.